tls: fix writeQueueSize prop, long write timeouts

apapirovski opened the pull request:
This PR addresses a long-standing issue of `_handle.writeQueueSize` in TLS not representing the actual write queue (https://github.com/nodejs/node/issues/15005), as well as the fact that long-lasting writes on both net & tls (http & https) will currently timeout if a socket timeout is set (https://github.com/nodejs/node/issues/15082). The latter has been reported as being in relation to the keep alive timeout but in reality it's just a broad issue that is the most obvious with keep-alive since it's much shorter at 5s (vs 120s).

- Make `writeQueueSize` represent the actual size of the write queue within the TLS socket.
-  Add tls test to confirm that `bufferSize` works as expected.

Second commit:

- Add `getWriteQueueSize` on `TLSWrap` which updates and returns queue size (net & tls).
- Make `_onTimeout` check whether an active write is ongoing and if so, call `_unrefTimer` rather than emitting a timeout event.
- Add http & https test that checks whether long-lasting (but active) writes timeout or can finish writing as expected.

Haven't worked in C++ in a very, very long time so would definitely appreciate any and all feedback there. 

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
http, https, net, test, tls

mscdex commented:
/cc @indutny

apapirovski commented:
Just a minor update to use `fixtures.readKey` instead. (Indirectly benefiting and learning from code-and-learn 👍)

/cc @mcollina

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10452/

mcollina commented:
Is updating this state have a performance impact? Have you tried to run our benchmarks? I think we can run them on CI if you do not have the spare CPU cycles (they take a long time).

apapirovski commented:
I ran all the benchmarks and didn't see any discernible difference at a 100 runs. The way it's handled, it only updates it once when DoWrite is called and then when it's finished (flushed). Similar to the way it's handled in regular net socket.

apapirovski commented:
Might need to refactor the test a bit as it's failing on SmartOS. Will get it done this weekend.

The rest seem fine.

apapirovski commented:
I've updated the tests to, hopefully, be less flaky. Would appreciate if we could get another CI started. Thanks!

joyeecheung commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/10495/

apapirovski commented:
Well, this might take a while. If anyone has any ideas on making these less timeout dependant, I'm all ears.

apapirovski commented:
Ok, I think this should work... ? I can run a stress test on the test on my local system now and it doesn't fail. Since it's still in `sequential` (and should stay there), I'm hoping that'll make it not fail on the slower systems.

joyeecheung commented:
Another CI: https://ci.nodejs.org/job/node-test-pull-request/10498/

apapirovski commented:
Well, this seems to succeed everywhere but SmartOS so I guess an improvement... Anyone have any suggestions by any chance? Going to try setting up a SmartOS box to test on...

apapirovski commented:
Ok, one last try. If this doesn't fix SmartOS then there's a deeper issue that might not have anything to do with the test itself. I can run this for an hour on my system (96 instances at a time) without any failures (previous version would fail like 2.5% of the time).

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10527/

mcollina commented on test/sequential/test-http-keep-alive-large-write.js:
```diff
@@ -0,0 +1,63 @@
+        setTimeout(
+          resume,
+          serverTimeout + offsetTimeout - (Date.now() - firstReceivedAt)
+        );
```
is this `setTimeout` needed?

apapirovski commented:
Oops, I accidentally deleted a few characters before committing the http version of the test so it was testing with readSize of 3333 instead of 333333. No wonder it failed. Could we run the CI just on SmartOS?

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10565/

apapirovski commented:
Thanks for starting it @mcollina. No luck this time either :( No clue if there's a bug in http on SmartOS related to writeQueueSize or if it's just the test. Almost done setting up a SmartOS box so I guess I'll find out...

apapirovski commented:
Having a SmartOS box really made this easy to fix. Should be good for another CI run now.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10584/

apapirovski commented:
Thanks! Looks like it worked this time (the SmartOS failure is unrelated, it's in `parallel/test-benchmark-string_decoder`). Now to get some reviews... :)

ping @indutny since you know the most about the TLS code.

apapirovski commented:
ping @addaleax, @BridgeAR & @refack — would someone, anyone please consider reviewing this?

addaleax commented on src/tls_wrap.h:
```diff
@@ -132,6 +132,7 @@ class TLSWrap : public AsyncWrap,
 
   AsyncWrap* GetAsyncWrap() override;
   bool IsIPCPipe() override;
+  void UpdateWriteQueueSize(int write_queue_size = 0);
```
It seems like `0` might be a valid length for the write queue size (now or at least conceptually/in the future?), so maybe this could be `-1` for indicating a fallback/default?

apapirovski replied:
The idea here is that it's harmful for it to be set to 0 if `BIO_pending(enc_out_)` is not actually 0 — it can then lead to breaking the logic in `net.js`.

addaleax replied:
@apapirovski As far as I understand it, `write_queue_size = 0` is used to say “use the value provided by `BIO_pending`”? And `BIO_pending` can return 0, right? In that case I’d say using `-1` would be a better value to communicate “please fall back to `BIO_pending`”

apapirovski replied:
Right but we want to disallow someone forcing through `UpdateWriteQueueSize(0)` because if they do that and BIO_pending is not actually 0 then they potentially just broke stuff.

addaleax replied:
> Right but we want to disallow someone forcing through `UpdateWriteQueueSize(0)` because if they do that and BIO_pending is not actually 0 then they potentially just broke stuff.

Okay, that makes a bit more sense, but could you explain why?

apapirovski replied:
`net.js` depends on the writeQueueSize being > 0 if the write hasn't been flushed (`BIO_pending` returns > 0). I haven't followed the trail all the way but in my testing any time that wasn't the case, it broke everything.

It's also the reason we have to set `writeQueueSize` to 1 for empty writes. It's possible all that logic could be further improved but it would require far more significant changes than this.

addaleax commented on src/tls_wrap.cc:
```diff
@@ -929,6 +929,16 @@ int TLSWrap::SelectSNIContextCallback(SSL* s, int* ad, void* arg) {
+
+  wrap->UpdateWriteQueueSize();
+  args.GetReturnValue().Set(
+    wrap->object()->Get(wrap->env()->write_queue_size_string()));
```
Can you use the overloads of `Get()` and `Set()` that take a context argument/return a `Maybe`? The others are deprecated

apapirovski replied:
Sorry @addaleax, I have very little familiarity with all of this. Do you mind elaborating? 😓

addaleax replied:
There are two different versions of `Get()` and `Set()` each, one where the first parameter is a `context` argument and one where it is omitted (which is the deprecated one). Maybe it’s easiest to just look for the strings `Get(context` and `Set(context` in the codebase, to see how the other versions of `Get` and `Set` are used?

apapirovski commented:
I'll just ping @nodejs/http & @nodejs/crypto in case they have any input on this.

I think this ideally needs to land in 8.x to fix the existing (fairly serious) bug around writes timing out but I don't want to introduce other issues in the process. In a perfect world this would get a substantial number of reviews.

bnoordhuis commented on lib/net.js:
```diff
@@ -400,6 +400,12 @@ Socket.prototype.setTimeout = function(msecs, callback) {
+  if (writeQueueSize > 0 &&
+      writeQueueSize !== this._handle.getWriteQueueSize()) {
+    this._unrefTimer();
+    return;
```
I don't understand why the 'timeout' event is suppressed here.

As well, can you add a comment explaining the checks?  They're going to look tautologically wrong to a casual reader ("when would the queue size not equal the queue size?")

apapirovski replied:
So the current bug this is dealing with is that once you had something off to the backing stream, it can be actively writing but the socket doesn't know about it and if the timeout happens, it'll just destroy itself and the write that's currently occurring.

We don't update `writeQueueSize` as the write is ongoing, we only set it at the very beginning. So when one uses `this._handle.getWriteQueueSize()` they get an updated value and compare it to what it was when the write started. If the value changed the timeout is wrong because we're still writing behind the scenes, if the value didn't change then this is likely an actual timeout.

Does that sort of make some sense?

apapirovski replied:
`this._handle.getWriteQueueSize` shouldn't run unless prevQueueSize > 0 hence the current format. (It's not terribly slow but we also don't want to run it on every single timeout.) I'll adjust some things & add comment to make it clearer what's going on.

Re: timeout reset, it does. When the write is done, it triggers 'Socket.prototype.afterWrite` which calls `_unrefTimer`.

apapirovski commented on src/stream_wrap.cc:
```diff
@@ -273,6 +274,17 @@ void LibuvStreamWrap::OnRead(uv_stream_t* handle,
+
+  wrap->UpdateWriteQueueSize();
+  args.GetReturnValue().Set(
+    wrap->object()->Get(wrap->env()->write_queue_size_string()));
```
I think changing it to return it is probably sensible. Will update.

apapirovski commented on src/tls_wrap.cc:
```diff
@@ -551,6 +553,15 @@ bool TLSWrap::IsClosing() {
 }
 
 
+void TLSWrap::UpdateWriteQueueSize(int write_queue_size) {
```
Yep, good point. I know some projects take issue with `uint32_t` so wasn't sure. I'll use `size_t`, if no objections.

apapirovski commented:
>  I have the feeling we set it (which is expensive-ish) more often than is necessary.

Currently it's only set at the beginning of the write and once it ends. It's not updated at all during the write. Without it we won't be able to easily prevent timeouts while writes are occurring.

refack commented on test/sequential/test-http-keep-alive-large-write.js:
```diff
@@ -0,0 +1,62 @@
+        firstReceivedAt = Date.now();
+      }
+      receivedBufferLength += buf.length;
+      if (receivedBufferLength >= writtenOffset) {
```
If I understand this correctly this checks if we received more than 50%? If that is right could you do:
```js
const isHalfRecived = receivedBufferLength >= writtenOffset;
```
And comment why we need 50%?

apapirovski replied:
Not quite. Basically, by the time the first `data` event fires there has been a certain amount of stuff written. We get that amount and then consume it, then we call `pause`. This allows us to mimic a slow client. Then we wait longer than the timeout, _onTimeout fires and checks whether it had a chance to write anything or the timeout is real. Then we continue and do this as many times as necessary until the data is all consumed.

refack replied:
Ment to delete this comment (figured out the `writeSizeOnFirstRead` thing).
But if you decide to keep the var name a comment would be great.

refack commented on test/sequential/test-http-keep-alive-large-write.js:
```diff
@@ -0,0 +1,62 @@
+let offsetTimeout = common.platformTimeout(100);
+let socketHandle;
+let writeSize = 3000000;
+let writtenOffset = 0;
```
If I understand this correctly this is used to check how much was written by the time we get the first `data` event. If that is correct could you rename this to something similar to `writeSizeOnFirstRead`

apapirovski replied:
It's a bit tricky to name this well. I might just add a comment. It represents the write amount of each cycle where the cycle is `write > read > timeout event`.

refack commented on test/sequential/test-http-keep-alive-large-write.js:
```diff
@@ -0,0 +1,62 @@
+        res.pause();
+        setTimeout(
+          resume,
+          serverTimeout + offsetTimeout - (Date.now() - firstReceivedAt)
```
Could you comment on why we need this formula to calculate the throttle time?

apapirovski replied:
We want to always wait after the `_onTimeout` fires but not give it a chance to fire twice. On slower systems (or if this were in parallel) the processing of `data` can sometimes take long enough that after several iterations of reading, it's offset enough that two timeouts get a chance to fire.

refack commented:
LGTM, but the tests are a little bit hard to follow. Would appropriate some variable renames, and some comments.

apapirovski commented:
I'm reasonably certain I've addressed everyone's comments. I've also added an explanation of the two timeout tests that breaks down exactly what's going on. Local `make test` is still happy, tests still pass under stress test.

Oh, and re: above, I ended up using `uint32_t` as there's an existing `Set()` that uses it and not one for `size_t`. I know I could cast but that seems strange since the only reason I'm returning that value in the first place is to use it in that `Set()`.

bnoordhuis commented on src/tls_wrap.cc:
```diff
@@ -912,6 +932,15 @@ int TLSWrap::SelectSNIContextCallback(SSL* s, int* ad, void* arg) {
+  TLSWrap* wrap;
+  ASSIGN_OR_RETURN_UNWRAP(&wrap, args.Holder());
+
+  uint32_t write_queue_size = wrap->UpdateWriteQueueSize();
```
Is it correct to call `UpdateWriteQueueSize()` here?  Since it updates the `.writeQueueSize` property, I think it would depend on the order of evaluation whether `handle.writeQueueSize === handle.getWriteQueueSize()`.

apapirovski replied:
Would the left side ever get evaluated after the right side? But also we store the `writeQueueSize` in a variable before comparing it.

```js
  const prevWriteQueueSize = this._handle.writeQueueSize;
  if (prevWriteQueueSize > 0 &&
      prevWriteQueueSize !== this._handle.getWriteQueueSize()) {
```

Ultimately, we need a way to update `writeQueueSize` here (for the next time _onTimeout runs) so I just wanted to avoid having to go in there twice.

bnoordhuis replied:
Then can you call it `.updateWriteQueueSize()`, to make it clear it has side effects?

> Would the left side ever get evaluated after the right side?

If you have to worry about such subtleties, I'd say it's time to, ah, _re-evaluate_ your code. :-)

apapirovski commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1019/

Just to be safe, since I'm changing timeout behaviour and fixing TLS bufferSize behaviour.

apapirovski commented:
A SmartOS bug crept back into the tests... :( Should be fixed so running again.

CI: https://ci.nodejs.org/job/node-test-pull-request/10834/

Also CITGM seems to be fine. One oddity with spdy on a single system but I tested it locally and not seeing the same issue. Will rerun again if the CI passes.

apapirovski commented:
CI failures unrelated. SmartOS passed.

CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1020/

apapirovski commented:
Ok, CITGM was good. 

ping @addaleax, @refack & @mcollina since some not insignificant changes have been made since your approval.

addaleax commented on lib/net.js:
```diff
@@ -397,6 +397,14 @@ Socket.prototype.setTimeout = function(msecs, callback) {
+  if (prevWriteQueueSize > 0 &&
+      prevWriteQueueSize !== this._handle.updateWriteQueueSize()) {
+    this._unrefTimer();
+    return;
```
Why unref? That doesn't quite match what I understood when I read "suppress the timeout"?

apapirovski replied:
It'll call `_unrefActive` so that it can come back and check this condition again in whatever the `_idleTimeout` is, if need be. Possible the comment is confusing but I'm obviously familiar with what it does so I'm a bad judge.

addaleax replied:
So … I might be wrong here, but judging from the naming I’d think `_unrefActive` only affects whether the timer handle is ref()ed or not, i.e. whether it is keeping the process alive or not … does that sound about right?

apapirovski replied:
Sorry, missed the notification — `_unrefActive` re-schedules a timer for a timer that's a part of the `unrefedLists`.

https://github.com/nodejs/node/blob/686e092202514dfc3cf93a463e9177a80dd42133/lib/timers.js#L141-L151

Not sure if that's a great explanation... ? Let me know if it isn't.

addaleax replied:
Yeah, don’t feel like you need to wait on anything from me! There’s nothing obviously wrong in this PR.

And yes, this should just be renamed…

apapirovski replied:
👍 Just being extra safe since this will go into LTS and is a fairly major change to something that's been broken for quite a while. :)

addaleax replied:
Is there a specific time frame? You can just add a `baking-for-lts` label to indicate that the releaser (likely @gibfahn) should let this sit a while before pulling it in, *if* you would feel more comfortable with that.

(On the other hand, if you feel comfortable not waiting any longer than usual, feel free to voice that as well!)

apapirovski replied:
Since it's a reasonably serious bug fix, it hasn't broken CITGM and no one else has expressed desire for having it bake, I think having it go in at the next opportunity should be fine. Thanks again for reviewing ❤️

apapirovski commented:
@indutny Thanks for the review! I did run it through CITGM and it seemed to make it through unscathed.

apapirovski commented:
Landed in https://github.com/nodejs/node/commit/aaf2a1c2264fffd6eea82fe3778f41ac046e6349 and https://github.com/nodejs/node/commit/a627c5fc136b3233959e8e1bff90ae4b2d0ec322.
